In [ ]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

In [1]:
!pip install llama-index
!pip install llama-index-core
!pip install llama-index-llms-anthropic llama-index-multi-modal-llms-anthropic
!pip install llama-index-embeddings-huggingface
!pip install llama-parse
!pip install bitsandbytes
!pip install accelerate faiss-cpu pdfplumber
!pip install langchain langchain_openai langchain_chroma langchain_community

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.4/15.4 MB 79.1 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 63.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 147.9/147.9 kB 10.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 130.8/130.8 kB 9.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 53.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 328.5/328.5 kB 22.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 50.5 MB/s eta 0:00:00
  Attempting uninstall: nltk
    Found existing installation: nltk 3.2.4
    Uninstalling nltk-3.2.4:
      Successfully uninstalled nltk-3.2.4
  Attempting uninstall: beautifulsoup4
    Found existing installation: beautifulsoup4 4.12.2
    Uninstalling beautifulsoup4-4.12.2:
      Successfully uninstalled beautifulsoup4-4.12.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed.

# **Initializing**

In [2]:
# llama-parse is async-first, running the async code in a notebook requires the use of nest_asyncio
import nest_asyncio
nest_asyncio.apply()
from openai import AzureOpenAI
import os
import numpy as np
import os
from pathlib import Path
from llama_index.readers.file import PDFReader
import json
from langchain_openai import AzureOpenAIEmbeddings,AzureChatOpenAI


# API access to llama-cloud
os.environ["LLAMA_CLOUD_API_KEY"] = ""

# Using Anthropic API for embeddings/LLMs
os.environ["AZURE_OPENAI_API_KEY"] = ""
os.environ["AZURE_OPENAI_ENDPOINT"] = ""

llm = AzureChatOpenAI(
    azure_deployment="",
    api_version="",
    temperature=0
)

embeddings_model = AzureOpenAIEmbeddings(
    model = '',
    openai_api_version="",
)

def get_query(messages):
    response = llm.invoke(messages)
    return response.content

def get_embedding(text, model="text-embedding3"):
   text = text.replace("\n", " ")
   return embeddings_model.embed_query(text)


In [3]:
get_query("hello")

'Hello! How can I assist you today?'

# Multimodal vision 

In [4]:
import torch
from PIL import Image
from transformers import AutoModelForCausalLM, AutoTokenizer

device = "cuda"

tokenizer = AutoTokenizer.from_pretrained("THUDM/glm-4v-9b", trust_remote_code=True)


model = AutoModelForCausalLM.from_pretrained(
    "THUDM/glm-4v-9b",
    torch_dtype=torch.float16,
    load_in_4bit=True,
    low_cpu_mem_usage=True,
    trust_remote_code=True
).eval()

# query = 'Analyze this graph' 
# image = Image.open("/kaggle/input/utilss/1.png").convert('RGB')
# inputs = tokenizer.apply_chat_template([{"role": "user", "image": image, "content": query}],
#                                        add_generation_prompt=True, tokenize=True, return_tensors="pt",
#                                        return_dict=True).to("cuda")  # chat mode

# gen_kwargs = {"max_length": 1024, "do_sample": True, "top_k": 1}
# with torch.no_grad():
#     outputs = model.generate(**inputs, **gen_kwargs)
#     outputs = outputs[:, inputs['input_ids'].shape[1]:]
#     print(tokenizer.decode(outputs[0]))

tokenizer_config.json:   0%|          | 0.00/3.22k [00:00<?, ?B/s]

tokenization_chatglm.py:   0%|          | 0.00/17.5k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/THUDM/glm-4v-9b:
- tokenization_chatglm.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


tokenizer.model:   0%|          | 0.00/2.62M [00:00<?, ?B/s]

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


config.json:   0%|          | 0.00/1.77k [00:00<?, ?B/s]

configuration_chatglm.py:   0%|          | 0.00/2.57k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/THUDM/glm-4v-9b:
- configuration_chatglm.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


modeling_chatglm.py:   0%|          | 0.00/68.6k [00:00<?, ?B/s]

visual.py:   0%|          | 0.00/6.79k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/THUDM/glm-4v-9b:
- visual.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.
A new version of the following files was downloaded from https://huggingface.co/THUDM/glm-4v-9b:
- modeling_chatglm.py
- visual.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.
The `load_in_4bit` and `load_in_8bit` arguments are deprecated and will be removed in the future versions. Please, pass a `BitsAndBytesConfig` object in `quantization_config` argument instead.


model.safetensors.index.json:   0%|          | 0.00/111k [00:00<?, ?B/s]

model-00001-of-00015.safetensors:   0%|          | 0.00/1.95G [00:00<?, ?B/s]

model-00002-of-00015.safetensors:   0%|          | 0.00/1.82G [00:00<?, ?B/s]

model-00003-of-00015.safetensors:   0%|          | 0.00/1.97G [00:00<?, ?B/s]

model-00004-of-00015.safetensors:   0%|          | 0.00/1.93G [00:00<?, ?B/s]

model-00005-of-00015.safetensors:   0%|          | 0.00/1.82G [00:00<?, ?B/s]

model-00006-of-00015.safetensors:   0%|          | 0.00/1.97G [00:00<?, ?B/s]

model-00007-of-00015.safetensors:   0%|          | 0.00/1.93G [00:00<?, ?B/s]

model-00008-of-00015.safetensors:   0%|          | 0.00/1.82G [00:00<?, ?B/s]

model-00009-of-00015.safetensors:   0%|          | 0.00/1.97G [00:00<?, ?B/s]

model-00010-of-00015.safetensors:   0%|          | 0.00/1.97G [00:00<?, ?B/s]

model-00011-of-00015.safetensors:   0%|          | 0.00/1.96G [00:00<?, ?B/s]

model-00012-of-00015.safetensors:   0%|          | 0.00/1.98G [00:00<?, ?B/s]

model-00013-of-00015.safetensors:   0%|          | 0.00/1.96G [00:00<?, ?B/s]

model-00014-of-00015.safetensors:   0%|          | 0.00/1.98G [00:00<?, ?B/s]

model-00015-of-00015.safetensors:   0%|          | 0.00/811M [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/15 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/205 [00:00<?, ?B/s]

# **Parsing the pdf**

# 1. Parse the text (using sentence splitter)

In [ ]:
from pathlib import Path
from llama_index.readers.file import PDFReader
from llama_index.core.response.notebook_utils import display_source_node
import json
from llama_index.core import Document
from llama_index.core.node_parser import SentenceSplitter
from llama_index.core.schema import IndexNode,NodeRelationship
from faiss import IndexFlatIP

loader = PDFReader()
docs0 = loader.load_data(file=Path("/kaggle/input/utilss/a.pdf"))

doc_text = "\n\n".join([d.get_content() for d in docs0])
docs = [Document(text=doc_text)]

In [ ]:
node_parser = SentenceSplitter(chunk_size=512)

text_base_nodes = node_parser.get_nodes_from_documents(docs)
# set node ids to be a constant
for idx, node in enumerate(text_base_nodes):
    node.id_ = f"node-{idx}"

sub_chunk_sizes = [64, 128, 256]
sub_node_parsers = [
    SentenceSplitter(chunk_size=c, chunk_overlap=20) for c in sub_chunk_sizes
]

text_all_nodes = []
for base_node in text_base_nodes:
    for n in sub_node_parsers:
        sub_nodes = n.get_nodes_from_documents([base_node])
        sub_inodes = [
            IndexNode.from_text_node(sn, base_node.node_id) for sn in sub_nodes
        ]
        text_all_nodes.extend(sub_inodes)

    # also add original node to node
    original_node = IndexNode.from_text_node(base_node, base_node.node_id)
    text_all_nodes.append(original_node)
    
text_all_nodes_dict = {n.node_id: n for n in text_all_nodes}

 # Step 1: Compute embeddings for all nodes
text_embeddings = []
text_node_ids = []
for node_id, node in text_all_nodes_dict.items():
    embedding = get_embedding(node.text) 
    text_embeddings.append(embedding)
    text_node_ids.append(node_id)

# Convert embeddings to numpy array
text_embeddings = np.array(text_embeddings)

# Initialize FAISS index
text_index = IndexFlatIP(text_embeddings.shape[1])
text_index.add(text_embeddings)

In [ ]:
from collections import Counter

def text_query_engine(query, text_all_nodes_dict, k=5):
    # Step 2: Compute embedding for the query string
    query_embedding = get_embedding(query)  # Replace with your actual embedding function

    # Search for top k similar nodes
    D, I = text_index.search(np.array([query_embedding]), k)

    # Step 3: Retrieve base nodes and count their occurrences
    text_base_node_ids = []
    for idx in I[0]:
        similar_node_id = text_node_ids[idx]
        similar_node = text_all_nodes_dict.get(similar_node_id)

        if similar_node and NodeRelationship.SOURCE in similar_node.relationships:
            base_node_id = similar_node.relationships[NodeRelationship.SOURCE].node_id
            text_base_node_ids.append(base_node_id)
        elif similar_node:
            text_base_node_ids.append(similar_node.node_id)  # If no SOURCE relationship, consider the node itself

    if text_base_node_ids:
        # Find the two most common base nodes among the top k similar nodes
        most_common_base_nodes = Counter(text_base_node_ids).most_common(2)
        top_base_nodes = [text_all_nodes_dict.get(node_id) for node_id, _ in most_common_base_nodes]
        l =[]
        for z in top_base_nodes:
            l.append(z.node_id)
        print(l)
        return top_base_nodes

Example usage:
query_string = "your query string"
base_nodes = text_query_engine(query_string, text_all_nodes_dict, k=3)
if base_nodes:
    print(f"Most Similar Base Node: {base_node.node_id for base_node in base_nodes}")
else:
    print(f"No similar base node found for query: '{query_string}'")


# 2. Get tables and images

In [5]:
from llama_index.core import Settings

# Settings.llm = llm
# Settings.embed_model = embeddings
from llama_parse import LlamaParse

parser = LlamaParse(verbose=True)
json_objs = parser.get_json_result("/kaggle/input/utilss/a.pdf")
json_list = json_objs[0]["pages"]
len(json_list)

Started parsing the file under job_id cac11eca-805a-464a-b900-5174f3dfd367


15

In [ ]:
from llama_index.core.schema import TextNode
from typing import List


def get_text_nodes(json_list: List[dict]):
    text_nodes = []
    for idx, page in enumerate(json_list):
        text_node = TextNode(
            text=page["text"],
            metadata={
                "page": page["page"]
            }
        )
        text_nodes.append(text_node)
    return text_nodes
text_nodes = get_text_nodes(json_list)


In [ ]:
# call get_images on parser, convert to ImageDocuments
!mkdir images

from llama_index.core.schema import ImageDocument
from llama_index.multi_modal_llms.anthropic import AnthropicMultiModal

def get_image_text_nodes(json_objs: List[dict]):
    """Extract out text from images using a multimodal model."""
    anthropic_mm_llm = AnthropicMultiModal(max_tokens=300)
    image_dicts = parser.get_images(json_objs, download_path="images")
    image_documents = []
    img_text_nodes = []
    
    for image_dict in image_dicts:
        image_doc = ImageDocument(image_path=image_dict["path"])
        image_doc = Image.open(image_doc.image_path).convert('RGB')

        
        
#         response = anthropic_mm_llm.complete(
#             prompt="Describe the images as an alternative text",
#             image_documents=[image_doc],
#         )
        
        
        prompt = 'Analyze the graph if any or return the explanation of the image'
        inputs = tokenizer.apply_chat_template([{"role": "user", "image": image_doc, "content": prompt}],
                                               add_generation_prompt=True, tokenize=True, return_tensors="pt",
                                               return_dict=True).to("cuda")  # chat mode

        gen_kwargs = {"max_length": 1024, "do_sample": True, "top_k": 1}
        
        with torch.no_grad():
            outputs = model.generate(**inputs, **gen_kwargs)
            outputs = outputs[:, inputs['input_ids'].shape[1]:]
            response = tokenizer.decode(outputs[0])
        
        
        
        text_node = TextNode(
            text=str(response),
            metadata={"path": image_dict["path"]}
        )
        img_text_nodes.append(text_node)
    return img_text_nodes

image_text_nodes = get_image_text_nodes(json_objs)


In [ ]:
import numpy as np

# Convert text nodes to embeddings
def convert_nodes_to_embeddings(text_nodes):
    embeddings = []
    for node in text_nodes:
        embedding = get_embedding(node.get_content())
        if embedding:
            embeddings.append((node, embedding))
    return embeddings

# Get the embeddings and corresponding nodes
node_embeddings = convert_nodes_to_embeddings(image_text_nodes)
nodes, embeddings = zip(*node_embeddings)
embeddings = np.array(embeddings)

In [ ]:
import faiss

# Define the dimension of your embeddings (for example, 768 for OpenAI embeddings)
embedding_dim = len(embeddings[0])

# Create a FAISS index
index = faiss.IndexFlatL2(embedding_dim)  # Use L2 distance for similarity search
index.add(embeddings)  # Add embeddings to the index

In [ ]:
def search_similar_nodes(query, k=2):
    # Get the embedding of the query text
    query_embedding = get_embedding(query)
    if query_embedding is None:
        return []

    query_embedding = np.array([query_embedding])

    # Search for the top k similar embeddings
    distances, indices = index.search(query_embedding, k)
    print(distances)
    
    # Retrieve the corresponding nodes
    similar_nodes = [nodes[idx] for idx in indices[0]]
    for z in similar_nodes:
        print(z.get_content())
    return similar_nodes

# Example usage

#for tables and images
# query_text = "Explain the transformers model step by step"
# top_k_similar_nodes = search_similar_nodes(query_text, k=3)

# for i, node in enumerate(top_k_similar_nodes):
#     print(f"Node {i+1}:\nContent: {node.get_content()}\n, id: {node.id_}")
    
#for texts
# base_node = text_query_engine(query_text, text_all_nodes_dict, k=3)
# if base_node:
#     print(f"Most Similar Base Node: {base_node.get_content()}")
# else:
#     print(f"No similar base node found for query: '{query_string}'")

In [ ]:
def create_prompt(query, similar_nodes,text_base_nodes):
    context = ""
    for text_base_node in text_base_nodes:
        if text_base_node is not None:
            context+="\n\n" + text_base_node.get_content()
#     context = context+"\n\n".join([node.get_content() for node in similar_nodes])
    prompt = f"Use the following context to answer the question:\n\n{context}\n\nQuestion: {query}"
    return context, prompt

def query_engine(query):
#     similar_nodes = search_similar_nodes(query)
    similar_nodes = None
    text_base_nodes = text_query_engine(query, text_all_nodes_dict, k=5)
    context, prompt = create_prompt(query, similar_nodes,text_base_nodes)
    
    messages = [
        {"role": "system", "content": prompt},
        {"role": "user", "content": query}
    ]
    
    response = get_query(messages)
    return context, response

In [ ]:
# Example usage
query_text = "What are some empirical evaluations of gated recurrent neural networks on sequence modeling?"
context,response = query_engine(query_text)
print(response)

In [ ]:
context

# Testing using ragas

In [56]:
!pip install pymupdf ragas

/opt/conda/lib/python3.10/pty.py:89: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  pid, fd = os.forkpty()


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.5/3.5 MB 44.9 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.7/15.7 MB 78.1 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 91.5/91.5 kB 7.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 71.1/71.1 kB 4.3 MB/s eta 0:00:00


In [57]:
from ragas.testset.generator import TestsetGenerator
from ragas.testset.evolutions import simple, reasoning, multi_context
from langchain_community.document_loaders.pdf import PyMuPDFLoader
loader = PyMuPDFLoader('/kaggle/input/utilss/a.pdf')
documents = loader.load()


# generator with openai models
generator_llm = llm
critic_llm = llm
embeddings = embeddings_model

generator = TestsetGenerator.from_langchain(
    generator_llm=generator_llm,
    critic_llm=critic_llm,
    embeddings=embeddings,
)

In [58]:
testset = generator.generate_with_langchain_docs(documents, test_size=10, distributions={simple: 0.5, reasoning: 0.25, multi_context: 0.25})


embedding nodes:   0%|          | 0/30 [00:00<?, ?it/s]

Generating:   0%|          | 0/10 [00:00<?, ?it/s]

auto merging

In [59]:
testset.to_pandas()
questions = testset.to_pandas()["question"].to_list()
ground_truth = testset.to_pandas()["ground_truth"].to_list()

In [60]:
from ragas.metrics import (
    answer_relevancy,
    faithfulness,
    context_recall,
    context_precision,
    answer_relevancy,
    answer_similarity
)

In [61]:
from datasets import Dataset

questions = testset.to_pandas()["question"].to_list()
ground_truth = testset.to_pandas()["ground_truth"].to_list()

data = {"question": [], "answer": [], "contexts": [], "ground_truth": ground_truth}

for query in questions:
    c,response = text_query_engine(query)
    
    if c:
        data["question"].append(query)
        data["answer"].append(response)
        data["contexts"].append([c])

    
dataset = Dataset.from_dict(data)

In [62]:
dataset.to_pandas()

question  \
0  What types of regularization are employed duri...   
1  What are some state-of-the-art approaches in s...   
2  What are the dominant sequence transduction mo...   
3  What types of regularization are employed duri...   
4  What are some notable papers and authors in th...   
5  What's the title of the paper on empirical eva...   
6  Which paper evaluates GRNNs for sequence model...   
7  How does the Transformer model use multi-head ...   
8  How do the Transformer model's BLEU scores com...   

                                              answer  \
0  We employ three types of regularization during...   
1  Recurrent neural networks, long short-term mem...   
2  The dominant sequence transduction models are ...   
3  We employ three types of regularization during...   
4  Some notable papers and authors in the field o...   
5  Empirical evaluation of gated recurrent neural...   
6  Junyoung Chung, Çaglar Gülçehre, Kyunghyun Ch...   
7  The Transformer model uses multi-head attentio...   
8  The Transformer model's BLEU scores are 28.4 f...   

                                            contexts  \
0  [\n\nandsemanticstructureofthesentences.\n5 Tr...   
1  [\n\ndetail.Nikidesigned,implemented,tunedande...   
2  [\n\nProvidedproperattributionisprovided,Googl...   
3  [\n\nandsemanticstructureofthesentences.\n5 Tr...   
4  [\n\nInformationProcessingSystems,(NIPS),2016....   
5  [\n\ndetail.Nikidesigned,implemented,tunedande...   
6  [\n\n[4] JianpengCheng,LiDong,andMirellaLapata...   
7  [\n\nmodel\nwefounditbeneficialtolinearlyproje...   
8  [\n\nandsemanticstructureofthesentences.\n5 Tr...   

                                        ground_truth  
0  We employ three types of regularization during...  
1  Recurrent neural networks, long short-term mem...  
2  The dominant sequence transduction models are ...  
3  We employ three types of regularization during...  
4  Some notable papers and authors in the field o...  
5  Empirical evaluation of gated recurrent neural...  
6  Empirical evaluation of gated recurrent neural...  
7  The Transformer model uses multi-head attentio...  
8  The Transformer model achieves better BLEU sco...

In [68]:
from ragas import evaluate

from ragas.metrics import (
    faithfulness,
    answer_relevancy,
    context_relevancy,
    context_recall,
    context_precision,
    answer_correctness,
    answer_similarity,
    context_precision
)

result = evaluate(
    dataset = dataset,
    llm = llm,
    embeddings = embeddings,
    metrics=[
        answer_relevancy,
        context_precision,
        answer_correctness,
        answer_similarity,
    ],
)

result.to_pandas()


Evaluating:   0%|          | 0/36 [00:00<?, ?it/s]

question  \
0  What types of regularization are employed duri...   
1  What are some state-of-the-art approaches in s...   
2  What are the dominant sequence transduction mo...   
3  What types of regularization are employed duri...   
4  What are some notable papers and authors in th...   
5  What's the title of the paper on empirical eva...   
6  Which paper evaluates GRNNs for sequence model...   
7  How does the Transformer model use multi-head ...   
8  How do the Transformer model's BLEU scores com...   

                                              answer  \
0  We employ three types of regularization during...   
1  Recurrent neural networks, long short-term mem...   
2  The dominant sequence transduction models are ...   
3  We employ three types of regularization during...   
4  Some notable papers and authors in the field o...   
5  Empirical evaluation of gated recurrent neural...   
6  Junyoung Chung, Çaglar Gülçehre, Kyunghyun Ch...   
7  The Transformer model uses multi-head attentio...   
8  The Transformer model's BLEU scores are 28.4 f...   

                                            contexts  \
0  [\n\nandsemanticstructureofthesentences.\n5 Tr...   
1  [\n\ndetail.Nikidesigned,implemented,tunedande...   
2  [\n\nProvidedproperattributionisprovided,Googl...   
3  [\n\nandsemanticstructureofthesentences.\n5 Tr...   
4  [\n\nInformationProcessingSystems,(NIPS),2016....   
5  [\n\ndetail.Nikidesigned,implemented,tunedande...   
6  [\n\n[4] JianpengCheng,LiDong,andMirellaLapata...   
7  [\n\nmodel\nwefounditbeneficialtolinearlyproje...   
8  [\n\nandsemanticstructureofthesentences.\n5 Tr...   

                                        ground_truth  answer_relevancy  \
0  We employ three types of regularization during...          0.000000   
1  Recurrent neural networks, long short-term mem...          0.889019   
2  The dominant sequence transduction models are ...          0.803173   
3  We employ three types of regularization during...          0.960276   
4  Some notable papers and authors in the field o...          0.999830   
5  Empirical evaluation of gated recurrent neural...          0.527985   
6  Empirical evaluation of gated recurrent neural...          0.668493   
7  The Transformer model uses multi-head attentio...          0.894807   
8  The Transformer model achieves better BLEU sco...          0.794582   

   context_precision  answer_correctness  answer_similarity  
0                1.0            0.556921           0.727685  
1                1.0            0.861890           0.909099  
2                1.0            0.742097           0.968388  
3                1.0            0.234640           0.938558  
4                1.0            0.511327           0.891463  
5                1.0            0.981322           0.925288  
6                1.0            0.172388           0.689552  
7                1.0            0.667827           0.906602  
8                1.0            0.721302           0.885210

In [64]:
result.to_pandas()['answer'][5]

'Empirical evaluation of gated recurrent neural networks for sequence modeling.'

In [69]:
print(result)


{'answer_relevancy': 0.7265, 'context_precision': 1.0000, 'answer_correctness': 0.6055, 'answer_similarity': 0.8713}


usng both and parent chunk 1024:
{'faithfulness': 0.7773, 'answer_relevancy': 0.8276, 'answer_correctness': 0.5453, 'answer_similarity': 0.6732}

reduced parent chunk and no images:
{'faithfulness': 0.9408, 'answer_relevancy': 0.8530, 'answer_correctness': 0.5791, 'answer_similarity': 0.7451, 'context_recall': 0.6833, 'context_relevancy': 0.0841}

{'faithfulness': 0.9248, 'context_precision': 0.8000, 'context_recall': 0.6833, 'context_relevancy': 0.0632}


full pipeline:
{'faithfulness': 0.7407, 'answer_relevancy': 0.7265, 'context_precision': 1.0000, 'answer_correctness': 0.6055, 'answer_similarity': 0.8713, 'context_recall': 0.9444, 'context_relevancy': 0.0383}

In [ ]:
0.17,1,0.95,s - 0.93, c - 0.61

# Using Recursive retriever


In [15]:
from langchain.retrievers import ParentDocumentRetriever, BM25Retriever, EnsembleRetriever
from langchain_core.prompts import ChatPromptTemplate, HumanMessagePromptTemplate, PromptTemplate
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_core.output_parsers import StrOutputParser
from langchain.vectorstores import Chroma
from langchain.storage import InMemoryStore
import pdfplumber
import os

In [74]:
class Document:
    def __init__(self, content, metadata=None):
        self.page_content = content
        self.metadata = metadata if metadata is not None else {}
"""
A class is like a blueprint for creating objects. 
An object is a collection of data (variables) and methods (functions) 
that act on the data. For example, a Document class can be used 
to create Document objects, each with its own content and metadata.
"""

# Load selected document

def get_image_description(image_path):
    print(image_path)
    image_doc = Image.open(image_path).convert('RGB')
#     image_doc = image
    prompt = 'Analyze the graph if any or return the explanation of the image in 200 words'
    inputs = tokenizer.apply_chat_template([{"role": "user", "image": image_doc, "content": prompt}],
                                           add_generation_prompt=True, tokenize=True, return_tensors="pt",
                                           return_dict=True).to("cuda")  # chat mode

    gen_kwargs = {"max_length": 1024, "do_sample": True, "top_k": 1}

    with torch.no_grad():
        outputs = model.generate(**inputs, **gen_kwargs)
        outputs = outputs[:, inputs['input_ids'].shape[1]:]
        response = tokenizer.decode(outputs[0])
        
    return response

        

def start_image_processing(page):
    images_in_page = page.images
    page_height = page.height
    i=1
    for image in images_in_page:
        image_bbox = (image['x0'], page_height - image['y1'], image['x1'], page_height - image['y0'])
        cropped_page = page.crop(image_bbox)
        image_obj = cropped_page.to_image(resolution=400)
        page_num = image['page_number']
        image_path = f"/kaggle/working/images_to_parse/{page_num}_{i}.png"
        image_obj.save(f"/kaggle/working/images_to_parse/{image['page_number']}_{i}.png")
        description = get_image_description(image_path)
        i+=1
    return description
    


def load_single_document(file_path):
    # Print the start of the document loading process
    print(f"Starting to load document from {file_path}")
    
    if os.path.exists(file_path) and file_path.endswith(".pdf"):
        # Print the filename of the PDF being processed
        filename = os.path.basename(file_path)
        print(f"Loading PDF file: {filename}")
        
        with pdfplumber.open(file_path) as pdf:
            full_text = []
            # Print the number of pages in the PDF
            print(f"Number of pages in PDF: {len(pdf.pages)}")
            
            for page_number, page in enumerate(pdf.pages, start=1):
                page_text = page.extract_text()
                
                if page.images:
                    print(page_number)
                    description = start_image_processing(page)
                    print(description)
                    page_text+=description
                    
                if page_text:
                    full_text.append(page_text)
                    
                # Print confirmation for each page processed
                print(f"Processed page {page_number}/{len(pdf.pages)}")
            
            # Combine all pages' text into one string for the document
            document_text = '\n'.join(full_text)
            # Create a Document object with the content and optional metadata
            document = Document(content=document_text, metadata={'filename': filename})
            # Print that the document has been successfully created
            print(f"Created document for {filename} with {len(full_text)} pages of text.")
    else:
        print(f"File not found or is not a PDF: {file_path}")
        return None
    
    return document

In [ ]:
# Specify the path to the PDF file
file_path = '/kaggle/input/utilss/a.pdf'
document = load_single_document(file_path) # Load the document
docs=[]
docs.append(document)

# from langchain_community.document_loaders import PyMuPDFLoader
# loader = PyMuPDFLoader("/kaggle/input/utilss/a.pdf")
# docs = loader.load()

Starting to load document from /kaggle/input/utilss/a.pdf
Loading PDF file: a.pdf
Number of pages in PDF: 15
Processed page 1/15
Processed page 2/15
3
/kaggle/working/images_to_parse/3_1.png


In [49]:
coll = vectorstore.get()  # dict_keys(['ids', 'embeddings', 'documents', 'metadatas'])

ids_to_del = []

for idx in range(len(coll['ids'])):

    id = coll['ids'][idx]
    metadata = coll['metadatas'][idx]
    ids_to_del.append(id)

vectorstore._collection.delete(ids_to_del)

NameError: name 'vectorstore' is not defined

In [50]:
# This text splitter is used to create the parent documents
parent_splitter = RecursiveCharacterTextSplitter(chunk_size=2000,chunk_overlap = 50)

# This text splitter is used to create the child documents
# It should create documents smaller than the parent
child_splitter = RecursiveCharacterTextSplitter(chunk_size=400, chunk_overlap = 20)

# The vectorstore to use to index the child chunks
vectorstore = Chroma(collection_name="return_split_parent_documents", embedding_function=embeddings_model)

# The storage layer for the parent documents
store = InMemoryStore()

retriever = ParentDocumentRetriever(
    vectorstore=vectorstore, 
    docstore=store, 
    child_splitter=child_splitter,
    parent_splitter=parent_splitter,
)

retriever.add_documents(docs)

In [51]:
vectorstore.similarity_search("what is  attention?")

[Document(metadata={'doc_id': '53a8eec7-fc9e-41f8-848b-80b663b7dd78', 'filename': 'a.pdf'}, page_content='Self-attention,sometimescalledintra-attentionisanattentionmechanismrelatingdifferentpositions\nofasinglesequenceinordertocomputearepresentationofthesequence. Self-attentionhasbeen\nusedsuccessfullyinavarietyoftasksincludingreadingcomprehension,abstractivesummarization,\ntextualentailmentandlearningtask-independentsentencerepresentations[4,27,28,22].'),
 Document(metadata={'doc_id': '84a90845-2618-4e14-a9e0-599a5fc338a9', 'filename': 'a.pdf'}, page_content='3.2 Attention\nAnattentionfunctioncanbedescribedasmappingaqueryandasetofkey-valuepairstoanoutput,\nwherethequery,keys,values,andoutputareallvectors. Theoutputiscomputedasaweightedsum\n3\nScaledDot-ProductAttention Multi-HeadAttention\nFigure 2: (left) Scaled Dot-Product Attention. (right) Multi-Head Attention consists of several\nattentionlayersrunninginparallel.'),
 Document(metadata={'doc_id': 'a807e82d-622c-4c9c-8137-ed89efebd

In [73]:
retriever.get_relevant_documents("explain the Scaled Dot-Product Attention mechanism")


[Document(metadata={'filename': 'a.pdf'}, page_content='3.2 Attention\nAnattentionfunctioncanbedescribedasmappingaqueryandasetofkey-valuepairstoanoutput,\nwherethequery,keys,values,andoutputareallvectors. Theoutputiscomputedasaweightedsum\n3\nScaledDot-ProductAttention Multi-HeadAttention\nFigure 2: (left) Scaled Dot-Product Attention. (right) Multi-Head Attention consists of several\nattentionlayersrunninginparallel.\nofthevalues,wheretheweightassignedtoeachvalueiscomputedbyacompatibilityfunctionofthe\nquerywiththecorrespondingkey.\n3.2.1 ScaledDot-ProductAttention\nWecallourparticularattention"ScaledDot-ProductAttention"(Figure2). Theinputconsistsof\nqueriesandkeysofdimensiond k,a√ndvaluesofdimensiond v. Wecomputethedotproductsofthe\nquerywithallkeys,divideeachby d ,andapplyasoftmaxfunctiontoobtaintheweightsonthe\nk\nvalues.\nInpractice,wecomputetheattentionfunctiononasetofqueriessimultaneously,packedtogether\nintoamatrixQ. ThekeysandvaluesarealsopackedtogetherintomatricesK andV. Wec

In [53]:
def text_query_engine(query):
    prompt_template = """Use the following pieces of context to answer the question at the end.
    If you don't know the answer, just say that you don't know, don't try to make up an answer.

    {context}

    Question: {question}
    Answer:"""
    PROMPT = PromptTemplate(
        template=prompt_template, input_variables=["context", "question"]
    )

    larger_chunk_relevant_docs =retriever.get_relevant_documents(query)

    content = ""
    for cont in larger_chunk_relevant_docs:
        content+='\n\n'+cont.page_content

    response = llm.invoke(input=PROMPT.format_prompt(
        context=larger_chunk_relevant_docs,
        question=query
    ).text).content
    
    return content,response
    

In [54]:
content,response = text_query_engine("what is attention?")

In [55]:
response

'An attention function can be described as mapping a query and a set of key-value pairs to an output, where the query, keys, values, and output are all vectors. The output is computed as a weighted sum of the values, where the weight assigned to each value is computed by a compatibility function of the query with the corresponding key.'

# Images and tables

In [6]:
rmdir "/kaggle/working/images_to_parse"

rmdir: failed to remove '/kaggle/working/images_to_parse': No such file or directory


In [29]:
import pdfplumber

!mkdir /kaggle/working/images_to_parse
pdf_obj = pdfplumber.open("/kaggle/input/utilss/a.pdf")

for page in pdf_obj.pages:
    i=1
    images_in_page = page.images
    page_height = page.height
    for image in images_in_page:
        image_bbox = (image['x0'], page_height - image['y1'], image['x1'], page_height - image['y0'])
        cropped_page = page.crop(image_bbox)
        image_obj = cropped_page.to_image(resolution=400)
        image_obj.save(f"/kaggle/working/images_to_parse/{image['page_number']}_{i}.png")
        i+=1

mkdir: cannot create directory '/kaggle/working/images_to_parse': File exists
3
4
4


In [65]:
page = pdf_obj.pages[10]
page.extract_table()

# ---------------------------------------------------------------------------------------------

# ----------------------------------------JUNK---------------------------------------------------

# -----------------------------------------------------------------------------------------------------

In [ ]:
# Split your document into big chunks
parent_splitter = RecursiveCharacterTextSplitter(chunk_size=5000, chunk_overlap=50)

# This text splitter is used to create the child documents. They should be small chunk size.
child_splitter = RecursiveCharacterTextSplitter(chunk_size=400, chunk_overlap=10)

# # Split it up
child_chunks = child_splitter.split_documents(docs)
print (f"Your document has been split into {len(child_chunks)} chunks")

In [ ]:
# The vectorstore to use to index the child chunks
child_vectorstore = Chroma(
    collection_name="parent_document_splits",
    embedding_function=embeddings_model
)

# The storage layer for the parent documents
parent_docstore = InMemoryStore()

retriever_parent = ParentDocumentRetriever(
    vectorstore=child_vectorstore, 
    docstore=parent_docstore,
    child_splitter=child_splitter,
    parent_splitter=parent_splitter)

retriever_parent.add_documents(docs)


num_parent_docs = len(retriever_parent.docstore.store.items())
num_child_docs = len(set(retriever_parent.vectorstore.get()['documents']))

print (f"You have {num_parent_docs} parent docs and {num_child_docs} child docs")

In [ ]:
# Define the human message prompt template
human_message_template = PromptTemplate.from_template(
    """
    You are an experienced assistant specializing in question-answering tasks. 
    Utilize the provided context to respond to the question. 
    Never provide an answer you are unsure about and ensure it is concise.
    Your answer must be comprehensive and contain all of the available details in the Context.
    \nQuestion: {question} \nContext: {context} \nAnswer:
    """
)

In [ ]:
def query_engine(query):
    # Create a HumanMessagePromptTemplate instance using the defined prompt template
    human_message_prompt_template = HumanMessagePromptTemplate(prompt=human_message_template)

    # Create the ChatPromptTemplate with the input variables and messages, excluding metadata
    chat_prompt_template = ChatPromptTemplate(
        input_variables=['context', 'question'],
        messages=[human_message_prompt_template]
    )
    gpt = AzureChatOpenAI(temperature=0, model_name="GPT35-turboA", api_key=os.getenv('AZURE_OPENAI_API_KEY'),openai_api_version="2024-02-01")

    rag_chain = chat_prompt_template | gpt | StrOutputParser()
    #question = 'Summary the provided context'
    question = query
    child_docs = child_vectorstore.similarity_search(question)
    parent_relevant_docs = retriever_parent.invoke(question)
    
    print (f"Child docs found: {len(child_docs)} ") 
    print (f"Parent docs retrieved: {len(parent_relevant_docs)}")
    
    context= ""
    if parent_relevant_docs:
        for x in parent_relevant_docs:
            context+=x.page_content+"\n\n"
    else:
        for x in child_docs:
            context+=x.page_content+"\n\n"
    
    generation = rag_chain.invoke({"context": context, "question": question})
    return context,generation
    